orderbook.js

```diff
@@ -0,0 +1,35 @@
+const reconcileOrder = (existingBook, incomingOrder) => {
+  const oppositeOrders = existingBook.filter(order => order.type !== incomingOrder.type)
+  let updatedBook = existingBook.filter(order => order.type === incomingOrder.type)
+
+  updatedBook = existingBook.length ? updatedBook.concat(fillOrAddOrder(oppositeOrders, incomingOrder))
+    : updatedBook.concat(incomingOrder)
+
+  return updatedBook
+}
+
+const fillOrAddOrder = (existing, incoming) => {
+  const index = incoming.type === 'sell'
+    ? existing.findIndex(order => order.price >= incoming.price)
+    : existing.findIndex(order => order.price <= incoming.price)
```

Comment:
This type of looping should be handled in reconcileOrder since it is flow of control. Those filters in reconcileOrder are unnecessary, you should be dealing with one existing order at a time. Break the problem down, don't try to solve it all in one go.